src/core/license.rs

```diff
@@ -401,66 +426,94 @@ impl BinaryBlob {
     }
 }
 
-#[allow(dead_code)]
-pub struct UpgradeLicense {
-    mac_data: Vec<u8>,
-    encrypted_license_data: Vec<u8>,
-}
-
-impl UpgradeLicense {
-    fn from_bytes(raw: &mut dyn Read) -> RdpResult<Self> {
-        let mut message = component![
-            "EncryptedLicenseInfo" => component![
-                "wBlobType" => U16::LE(0),
-                "wBlobLen" => DynOption::new(U16::LE(0), | size | MessageOption::Size("blobData".to_string(), size.inner() as usize)),
-                "blobData" => Vec::<u8>::new()
-            ],
-            "MACData" => vec![0_u8; 16]
-        ];
-        message.read(raw)?;
-
-        let encrypted_license_data =
-            cast!(DataType::Slice, message["EncryptedLicenseInfo"])?.to_vec();
-        let mac_data = cast!(DataType::Slice, message["MACData"])?.to_vec();
-
-        Ok(Self {
-            encrypted_license_data,
-            mac_data,
-        })
-    }
-
-    // TODO(zmb3): remove this exception
-    #[allow(dead_code)]
-    fn decrypted_license(&self, session_encryption_data: &SessionEncryptionData) -> Vec<u8> {
-        session_encryption_data.decrypt_message(&self.encrypted_license_data)
-    }
-}
+/// ServerUpgradeLicense is sent from server to client if the client
+/// presents an existing license and the server determines that this
+/// license should be upgraded.
+///
+/// The structure and content of this message is identical to the
+/// ServerNewLicense message.
+///
+/// See MS-RDPELE section 2.2.2.6.
+type ServerUpgradeLicense = ServerNewLicense;
 
 /// License data that has been obtained from the sever
 #[allow(dead_code)]
 struct License {
-    data: Vec<u8>,
+    // The version, scope, company name, and product ID fields are used
+    // by the client to index the licenses in the client's license store.
+    version_major: u16,
+    version_minor: u16,
+    scope: String,
+    company_name: String,
+    product_id: String,
+
+    // In theory, this is a DER-encoded X.509 certificate that has some
+    // helpful data in its certificate extensions:
+    //
+    // https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpele/99e6bdc5-4517-45a3-be72-a254e8389546
+    //
+    // In practice, something about this cert is non-standard: both Go's crypto/x509 package
+    // and Rust's x509_parser crate choke when attempting to parse it.
+    // (Note that `openssl asn1parse -inform DER` does work)
```

Comment:
Wasn't `x509` that worked while `asn1parse` choked like everything else?
```suggestion
    // (Note that `openssl x509 -inform DER` does work)
```

Reply:
Not anymore - I double checked the data and now it's the other way around 🤷